### DISCUSSION TITLE
How to sign in to github from the terminal?

### DISCUSSION THREAD

**Discussion by elkhot** [community/community#55691] [Other Feature Feedback, Questions, & Ideas] [answered]
### Select Topic Area

Question

### Body

I was trying to upload 6 (pdb) files 2.1 MB each, I do not why I managed to upload three of them through the website, then I tried to access the repository from my computer (I have ubuntu 22.04) I can not.
Signing in falls using the password, tooken or SSH

**Comment by jge162** [comment]
You need a token to push files via linux command line. Have you setup the virtual environment yet?

**Reply by elkhot** [reply]
I got this error:Password for 'https://token@github.com': 
To https://github.com/elkhot/Boltorn-4G_MD.git
 ! [rejected]        main -> main (non-fast-forward)
error: failed to push some refs to 'https://github.com/elkhot/Boltorn-4G_MD.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.

**Reply by airtower-luna** [reply, marked as answer]
@jge162, please don't tell people to put tokens into their URL without mentioning the security risks involved. :confused: Also the format in the picture doesn't work (it puts the token as username, not password, as evidenced by the password prompt), and either way (security problems aside) it's also rather inconvenient to have to update the remote URLs for all your repositories when the token expires. :sweat_smile: 

@elkhot, please check [Git Tools - Credential Storage](https://git-scm.com/book/en/v2/Git-Tools-Credential-Storage) for how to store your token without putting it into the URL. I'd use a password manager to store the token long-term, and enable the `cache` credential helper so I don't have to copy & paste it all the time. Or use SSH (that requires changing the URL from HTTPS to SSH).

The error you're getting now has nothing to do with authentication, it simply means your remote (the repository on GitHub) has commits that aren't in your local branch. What to do depends on whether you want to keep those:

* If you want to keep them, merge them into your local branch first (that's what the `git pull` hint is about).
* If you want to *permanently discard* them, you can force-push. If you do, there's no getting them back, so double-check. :warning:

**Reply by jge162** [reply]
Thanks for the info I’m still learning too @airtower-luna. I Did not know about the git credentials. Thanks ☺️
